server/src/main/java/org/apache/druid/messages/server/Outbox.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.messages.server;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.druid.messages.MessageBatch;
+import org.apache.druid.messages.client.MessageRelay;
+
+/**
+ * An outbox for messages sent from servers to clients.
```

Comment:
Question: Is outbox a batching mechanism, or also a way to order through the incoming messages?

Reply:
It does do batching but it's also order-preserving. I added a comment about that.